pyomo/core/base/component.py

```diff
@@ -438,19 +438,21 @@ def to_string(self, verbose=None, labeler=None, smap=None, compute_values=False)
                 pass
         return self.name
 
-    def getname(self, fully_qualified=False, name_buffer=None):
+    def getname(self, fully_qualified=False, name_buffer=None, relative_to=None):
         """
         Returns the component name associated with this object.
 
         Arguments:
             fully_qualified     Generate full name from nested block names
             name_buffer         Can be used to optimize iterative name
                                     generation (using a dictionary)
+            relative_to         When generating a fully qualified name,
+                                    stop at this block.
         """
         if fully_qualified:
             pb = self.parent_block()
-            if pb is not None and pb is not self.model():
-                ans = pb.getname(fully_qualified, name_buffer) \
+            if pb is not None and pb is not self.model() and (relative_to is None or relative_to is not pb):
+                ans = pb.getname(fully_qualified, name_buffer, relative_to) \
```

Comment:
This is technically redundant.
```
if pb is not None and pb is not self.model() and pb is not relative_to:
```
is sufficient as the first clause becomes redundant when relative_to is None.  Now that I think about it, I think a better (clearer) approach would be to make this code:
```
pb = self.parent_block()
if relative_to is None:
    relative_to = self.model()
if pb is not None and pb is not relative_to:
    ans = pb.getname(fully_qualified, name_buffer, relative_to) + "." + self._name
```
This approach explains that the name is relative to the model unless explicitly told otherwise.  It also eliminates a redundant calls to `model()` in deeply-nested models.